File: src/main/java/chapter2/dao/UserDao.java
package chapter2.dao;

import model.User;
import org.springframework.dao.EmptyResultDataAccessException;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class UserDao {

    private final DataSource dataSource;

    public UserDao(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    public void add(User user) throws SQLException {
        Connection connection = dataSource.getConnection();
        PreparedStatement preparedStatement = connection.prepareStatement("insert into users(id, name, password) values(?,?,?)");
        preparedStatement.setString(1, user.getId());
        preparedStatement.setString(2, user.getName());
        preparedStatement.setString(3, user.getPassword());

        preparedStatement.execute();

        preparedStatement.close();
        connection.close();
    }

    public User get(String id) throws SQLException {
        Connection connection = dataSource.getConnection();
        PreparedStatement preparedStatement = connection.prepareStatement("select * from users where id = ?");
        preparedStatement.setString(1, id);

        ResultSet resultSet = preparedStatement.executeQuery();

        User user = null;
        if (resultSet.next()) {
            user = new User(resultSet.getString("id"),
                    resultSet.getString("name"),
                    resultSet.getString("password"));
        }

        resultSet.close();
        preparedStatement.close();
        connection.close();

        if (user == null) throw new EmptyResultDataAccessException(1);

        return user;
    }

    public void deleteAll() throws SQLException {
        Connection connection = dataSource.getConnection();

        PreparedStatement preparedStatement = connection.prepareStatement("delete from users");
        preparedStatement.executeUpdate();

        preparedStatement.close();
        connection.close();
    }

    public int getCount() throws SQLException {
        Connection connection = dataSource.getConnection();

        PreparedStatement preparedStatement = connection.prepareStatement("select count(*) from users");

        ResultSet resultSet = preparedStatement.executeQuery();
        resultSet.next();
        int count = resultSet.getInt(1);

        resultSet.close();
        preparedStatement.close();
        connection.close();

        return count;
    }
}

File: spring6/src/main/kotlin/com/bifos/toby/spring6/chapter1/extension_dao/UserDao.kt
package com.bifos.toby.spring6.chapter1.extension_dao

import com.bifos.toby.spring6.chapter1.common.User

class UserDao(
    private val connectionMaker: ConnectionMaker,
) {
    fun add(user: User) {
        val connection = connectionMaker.makeNewConnection()
        val preparedStatement = connection.prepareStatement("insert into users(id, name, password) values(?, ?, ?)")
        preparedStatement.setString(1, user.id)
        preparedStatement.setString(2, user.name)
        preparedStatement.setString(3, user.password)

        preparedStatement.executeUpdate()

        preparedStatement.close()
        connection.close()
    }

    fun get(id: String): User {
        val connection = connectionMaker.makeNewConnection()

        val preparedStatement = connection.prepareStatement("select * from users where id = ?")
        preparedStatement.setString(1, id)

        val resultSet = preparedStatement.executeQuery()
        resultSet.next()
        val user = User(resultSet.getString("id"), resultSet.getString("name"), resultSet.getString("password"))

        resultSet.close()
        preparedStatement.close()
        connection.close()

        return user
    }
}

fun main() {
    val simpleConnectionMaker = SimpleConnectionMaker()
    val dao = UserDao(simpleConnectionMaker)

    val user = User("bifos", "김병태", "1234")
    dao.add(user)

    println("${user.id} 등록 성공")

    val user2 = dao.get(user.id)
    println(user2.name)
    println(user2.password)
    println("${user2.id} 조회 성공")
}